README.md
---
layout: default
title: Home
nav_order: 1
permalink: /
---

![header](docs/img/audioMotion-header.png "audioMotion - high-resolution real-time spectrum analyzer and music player")

[![GitHub release](https://img.shields.io/github/release/hvianna/audioMotion.js.svg) ![GitHub Release Date](https://img.shields.io/github/release-date/hvianna/audioMotion.js.svg)](https://github.com/hvianna/audioMotion.js/releases/latest)

♪♫ **See** your music with **audioMotion**! ♪♫

**audioMotion.js** is a high-resolution real-time audio spectrum analyzer and full-featured music player written in JavaScript.

The spectrum analyzer module is also available as a [standalone project](https://github.com/hvianna/audioMotion-analyzer) and a zero-dependency [npm package](https://www.npmjs.com/package/audiomotion-analyzer) you can use in your own JavaScript projects!

## Features

* High-resolution (retina / HiDPI ready) real-time audio spectrum analyzer with fullscreen support
* Logarithmic frequency scale with customizable range
* Visualize discrete frequencies with full FFT resolution, or octave bands based on the equal tempered scale
* Optional effects: vintage LEDs, luminance bars, reflection, radial visualization
* 17 beautiful color gradients
* Customizable Web Audio API parameters (FFT size, sensitivity and time-smoothing constant)
* HTML5 audio player supports MP3, AAC (m4a), OGG and FLAC files (support may vary depending on browser)
* Visual file explorer and play queue with drag-and-drop functionality
* Support for M3U playlists (`m3u` and `m3u8` file extensions)
* Visualize audio from your microphone (or "stereo mix", if your soundcard supports it)
* Lightweight custom file server — also works on standard web servers with minimal configuration
* Client runs on any modern browser (tested on Chrome, Firefox, Opera and Safari)

## Online demo

[**DEMO SITE** ▶ audioMotion.me/public](https://audiomotion.me/public/)

The demo allows you to play audio files from your PC, load songs and streams from remote URLs and use the microphone input.
The file explorer is not available, since there are no media files hosted on the demo server.

Also check out our [▶ **YouTube channel**](https://channel.audiomotion.io) for videos of some cool songs visualized with audioMotion!

## Download

Download the latest version from the [**releases page**](https://github.com/hvianna/audioMotion.js/releases/latest). Portable binaries are available for Windows, Linux and macOS.

Launch audioMotion and you'll be asked for the path to your music folder. Only files under that folder will be accessible to the player.

You can also use the command line argument `-m` to set the music folder when launching audioMotion:

```
audioMotion -m c:\users\john\music
```

This will start the server and should also launch the client in your browser.

The complete command line options are:

```
-m <path> : path to music directory
-p <port> : change server listening port (default is 8000)
-s        : start server only (do not launch client)
-e        : allow external connections (by default, only localhost)
```

⚠ **WARNING:**<br>
Please be aware that the `-e` flag will expose the contents of the mounted folder to anyone in your network (and potentially to the entire internet!) &mdash; use it only if you're in a trusted network and behind a firewall!

## Documentation

+ [User interface](docs/user-interface.md)
+ [Server modes](docs/server.md)
+ [Building audioMotion](docs/building.md)
+ [Legacy features](docs/legacy.md)

For full documentation summary, see the [docs folder](docs/).

## Screenshots

Discrete frequencies mode. User interface showing the file explorer and play queue.
![screenshot1](docs/img/screenshot1.png "Discrete frequencies mode. User interface showing the file explorer and play queue.")

1/4th-octave bands with LED effect, extra wide bar spacing and on-screen song information. User interface showing the settings panel.
![screenshot2](docs/img/screenshot2.png "1/4th-octave bands mode, extra wide bar spacing, with LED effect and on-screen song information. User interface showing the settings panel.")

Fullscreen 1/6th-octave bands, regular bar spacing, 30Hz-16KHz, "Outrun" gradient, Reflex effect and complete on-screen information.
![screenshot3](docs/img/screenshot3.png "Fullscreen 1/6th-octave bands, Regular bar spacing, 30Hz-16KHz, 'Outrun' gradient, Reflex effect and complete on-screen information")

Fullscreen Line graph, 20Hz-22KHz, frequency and level scales on, "Rainbow" gradient.
![screenshot6](docs/img/screenshot6.png "Fullscreen Line graph mode, 20Hz-22KHz, frequency and level scales on, 'Rainbow' gradient")

Fullscreen 1/8th-octave bands with Luminance bars effect, 30Hz-16KHz, frequency scale on, "Tie Dye" gradient.
![screenshot5](docs/img/screenshot5.png "Fullscreen 1/8th-octave bands mode with Luminance bars effect, 30Hz-16KHz, frequency scale on, 'Tie Dye' gradient")

Fullscreen Radial 1/6th-octave bands with frequency scale on, "Apple ][" gradient.
![screenshot7](docs/img/screenshot7.png "Fullscreen Radial 1/6th-octave bands mode with frequency scale on, 'Apple ][' gradient")

## Credits and acknowledgments

**audioMotion** was largely inspired by [Soniq Viewer for iOS](https://itunes.apple.com/us/app/soniq-viewer/id448343005), by Yuji Koike.

### JavaScript libraries

Some of audioMotion's great features are provided by these awesome open-source libraries:

* [express](https://www.npmjs.com/package/express) - fast, unopinionated, minimalist web framework for node
* [music-metadata-browser](https://www.npmjs.com/package/music-metadata-browser) - stream and file based music metadata parser for the browser
* [notie](https://www.npmjs.com/package/notie) - clean and simple notification, input, and selection suite for javascript, with no dependencies
* [open](https://www.npmjs.com/package/open) - open stuff like URLs, files, executables
* [pkg](https://www.npmjs.com/package/pkg) - single-command Node.js binary compiler
* [readline-sync](https://www.npmjs.com/package/readline-sync) - synchronous Readline for interactively requesting user input via console
* [scrollIntoViewIfNeeded 4 everyone](https://gist.github.com/hsablonniere/2581101) - polyfill for non-standard scrollIntoViewIfNeeded() method
* [sortablejs](https://www.npmjs.com/package/sortablejs) - JavaScript library for reorderable drag-and-drop lists
* [webpack](https://www.npmjs.com/package/webpack) - JavaScript module bundler for the browser
  * [mini-css-extract-plugin](https://www.npmjs.com/package/mini-css-extract-plugin) - extracts CSS into separate files
  * [optimize-css-assets-webpack-plugin](https://www.npmjs.com/package/optimize-css-assets-webpack-plugin) - a webpack plugin to optimize / minimize CSS assets
  * [terser-webpack-plugin](https://www.npmjs.com/package/terser-webpack-plugin) - JavaScript compressor

### Graphic resources

* Icons provided by [icons8](https://icons8.com) licensed under [Creative Commons Attribution-NoDerivs 3.0 Unported](https://creativecommons.org/licenses/by-nd/3.0/)
* Gradients [*Candy*](https://gradienthunt.com/gradient/172), [*Miami*](https://gradienthunt.com/gradient/950), [*Outrun*](https://gradienthunt.com/gradient/317) and [*Tie Dye*](https://gradienthunt.com/gradient/969) from [Gradient Hunt](https://gradienthunt.com)
* Gradients *Pacific Dream*, *Shahabi*, *Summer* and *Sunset* from [uiGradients](https://uigradients.com)
* The *Apple ][* gradient was created based on [this post](https://creativemarket.com/blog/6-famous-logos-with-great-color-schemes) by Creative Market
* Gradients used in the UI buttons are from [ColorZilla's Ultimate CSS Gradient Generator](http://www.colorzilla.com/gradient-editor/)
* The audioMotion logo uses the [Orbitron font](https://fonts.google.com/specimen/Orbitron) by Matt McInerney

### References

These online references were invaluable for the development and improvement of audioMotion:

* [HTML Canvas Reference @W3Schools](https://www.w3schools.com/tags/ref_canvas.asp)
* [Web Audio API documentation @MDN](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API)
* [Web Audio API Specification](https://webaudio.github.io/web-audio-api/)
* [What does the FFT data in the Web Audio API correspond to?](https://stackoverflow.com/a/14789992/2370385)
* [HTML5 check if audio is playing?](https://stackoverflow.com/a/46117824/2370385)
* [Unlocking Web Audio — the smarter way](https://hackernoon.com/unlocking-web-audio-the-smarter-way-8858218c0e09)
* [Equations for equal-tempered scale frequencies](http://pages.mtu.edu/~suits/NoteFreqCalcs.html)
* [Efficient case-insensitive sorting in JavaScript](https://stackoverflow.com/a/40390844/2370385)
* [Fastest way to find the index of a child node in parent](https://stackoverflow.com/a/13657635/2370385)
* [Custom CSS media element selectors for Webkit browsers](https://stackoverflow.com/a/33948255/2370385)
* Pure CSS modal windows based on [this pen](https://codepen.io/timothylong/pen/HhAer) by Timothy Long and [this project](https://drublic.github.io/css-modal/) by Hans Christian Reinl
* [CSS3 inset text-shadow](https://codepen.io/adambundy/pen/HtmaK) by Adam Bundy
* [CSS slider switches](https://codepen.io/billyysea/pen/CHmiE) by Billy

## License

audioMotion.js copyright (c) 2018-2020 Henrique Avila Vianna<br>
Licensed under the [GNU Affero General Public License, version 3 or later](https://www.gnu.org/licenses/agpl.html).

Changelog.md
---
layout: default
title: Changelog
nav_order: 3
---

# Changelog

## version NEXT

### Fixed:

+ Clicks on settings switches not being properly detected.


## version 20.9 :mask:

This is a minor update to address two bugs:

+ A day-one design flaw which connected the microphone output to the speakers, generating feedback loops;
+ An unexpected error message when trying to load a playlist with an empty value in the playlist selection.


## version 20.8 :mask:

### Added:

+ New [Radial visualization](docs/user-interface.md#mode-switches) for all modes;
+ Option to display [level (dB) scale](docs/user-interface.md#scale) on vertical axis;
+ New [**Demo** preset](docs/user-interface.md#preset).

### Changed:

+ Improved the background image [**Pulse**](docs/user-interface.md#image-fit) effect to look more synced regardless of music style;
+ Any image located in the song's folder can now be used as album cover when a picture is not found in the song metadata (see the [documentation](docs/user-interface.md#background) for filename precedence).

### Fixed:

+ Audio files with uppercase extensions not recognized when audioMotion was running on a standard web server.


## version 20.6 :mask:

### Added:

+ **Album cover image** retrieved from the song metadata or from a file named *cover* or *folder* (.jpg|png|gif|bmp) inside each folder
is now shown in the file explorer background, the on-screen song information and, optionally, in the analyzer background;

+ New [Background](docs/user-interface.md#background), [Image Fit](docs/user-interface.md#image-fit) and [Image Dim](docs/user-interface.md#image-dim) settings.

### Changed:

+ Slightly increased the opacity of the image reflection when Reflex is On.

### Removed:

+ The **NO BG** switch has been replaced by the new Background setting.

### Fixed:

+ Linux binary built with [latest pkg version](https://github.com/zeit/pkg/pull/751#issuecomment-626363292) can now open the browser automatically;
+ A typo that could cause "Unexpected null" errors on some browsers.


## version 20.4 :mask:

### Added:

+ New **Line graph** visualization mode, with customizable line width and fill opacity;
+ New **Reflex** effect;
+ New **Config panel** where you can:
  + Enable/disable visualization modes and gradients;
  + Select options affected by random mode;
  + Customize sensitivity presets (low, normal and high);
+ Customizable spacing between bars in octave bands modes;
+ Visualization mode can now be randomized on a time interval.

### Changed:

+ The **Area fill** mode has been renamed to **Area graph**;
+ The default spacing between bars in octave bands modes has been increased a bit - set the **Bar spacing** option to **Legacy** for the old look;
+ Slightly improved vertical usage of canvas when the LED effect is active (removed the black line at the bottom of the screen);
+ UI improvements.

### Fixed:

+ Track change sometimes not triggering random mode / auto gradient;
+ Case-insensitive sorting of directory listing when using a standard web server.


## version 19.12

:tada: Celebrating audioMotion's first anniversary! :confetti_ball:

### Added:

+ New **Area fill** visualization mode, which uses the same full-frequency data of the *discrete frequencies* mode, but displays a bright, colorful filled shape;
+ New luminance bars effect (**LUMI** switch) for octave bands modes, which always display full-height bars and vary their luminance instead, according to each band amplitude;
+ New option to select a random visualization mode on every track change (**RAND** switch); it will also select a random gradient, if the AUTO switch is active - great for parties!

### Changed:

+ Improved the look of bars at lower frequencies, especially for 1/12th and 1/24th octave bands modes;
+ Minor tweak to the **Rainbow** gradient to make cyan and blue shades a little more balanced;
+ Auto gradient and the new random mode now trigger on initial playback and previous track skip as well, not only on next track skip;
+ Shortcut keys changes:
  + Shuffle shortcut changed to **"E"** key;
  + **"U"** key reassigned to toggle the new luminance bars effect;
  + **"A"** key now also toggles random visualization mode in combination with auto gradient (three stages);
+ Added shortcut key hints to the player controls;
+ Improved display of feedback messages for keyboard controls;
+ Updated npm packages.

### Fixed:

+ Addressed an error on server startup when running the executable file on Linux systems (related to [an issue with pkg](https://github.com/zeit/pkg/issues/731)).


## version 19.10

:sunglasses: audioMotion's graphic spectrum analyzer is now available as a [standalone project](https://github.com/hvianna/audioMotion-analyzer) and a zero-dependency [npm package](https://www.npmjs.com/package/audiomotion-analyzer) you can use in your own JavaScript projects! :tada:

### Added:

+ 1/4th and 1/8th octave bands visualization modes;
+ Load files from remote URLs and manage playlists when running audioMotion in file mode (no server);

### Changed:

+ Double-clicking a playlist in the file explorer now correctly starts playing it, if player is idle;
+ Player automatically skips to next track on playback errors, e.g. file not found or format not supported;
+ Number of songs added to the playlist is now shown in a notification, instead of logged to the console;
+ Improved metadata fetching for streams and time display for longer audio files;
+ Improved error handling on file server;
+ Slightly improved gapless playback;
+ Updated npm packages.


## version 19.7

### Added:

+ :fire: File explorer for easy navigation through your music files and folders;
+ Custom file server written in node.js with portable binaries available for Windows, Linux and macOS - no setup required; :tada:
+ Add, reorder and remove songs from the play queue, with drag-and-drop support;
+ Song info now shows actual metadata read from music files;

src/server.js
/**
 * audioMotion.js custom file server
 * https://github.com/hvianna/audioMotion.js
 * Copyright (C) 2019-2020 Henrique Vianna <hvianna@gmail.com>
 */

const _VERSION = '20.9';

const serverSignature = `audioMotion.js server ver. ${_VERSION}`

const fs = require('fs')
const path = require('path')
const express = require('express')
const open = require('open')
const readlineSync = require('readline-sync')
const semver = require('semver')

var port = 8000
var host = 'localhost'
var launchClient = true
var musicPath = ''

function showHelp() {
	console.log( `
	Usage:

	audioMotion -m "${ process.platform == 'win32' ? 'c:\\users\\john\\music' : '/home/john/music' }"

	-m <path> : path to music directory
	-p <port> : change server listening port (default: ${port})
	-s        : start server only (do not launch client)
	-e        : allow external connections (by default, only localhost)

	` )
}

function getDir( directoryPath, showHidden = false ) {
	let dirs = [],
		files = [],
		entries = [];

	try {
		entries = fs.readdirSync( path.normalize( directoryPath ), { withFileTypes: true } );
	}
	catch( e ) {
		return false;
	}

	entries.forEach( entry => {
		if ( entry.name[0] != '.' || showHidden ) {
			if ( entry.isDirectory() )
				dirs.push( entry.name );
			else if ( entry.isFile() )
				files.push( entry.name );
		}
	});

	let collator = new Intl.Collator(); // for case-insensitive string sorting
	return { dirs: dirs.sort( collator.compare ), files: files.sort( collator.compare ) }
}

// helper function - find image files with given pattern in an array of filenames
function findImg( arr, pattern ) {
	const regexp = new RegExp( `${pattern}.*\\.(jpg|jpeg|png|gif|bmp)$`, 'i' );
	return arr.find( el => el.match( regexp ) );
}

/* main */

console.log( `\n${serverSignature}` );

if ( ! semver.gte( process.version, '10.10.0' ) ) {
	console.log( `\n\nERROR: the minimum required version of node.js is v10.10.0 and you're running ${process.version}` );
	process.exit(0);
}

// processes command line arguments
process.argv = process.argv.slice(2)
process.argv.forEach( ( arg, index ) => {
	if ( arg == '-h' || arg == '--help' ) {
		showHelp()
		process.exit(0)
	}
	else if ( arg == '-m' && process.argv[ index + 1 ] )
		musicPath = path.normalize( process.argv[ index + 1 ] )
	else if ( arg == '-p' && process.argv[ index + 1 ] > 0 )
		port = process.argv[ index + 1 ]
	else if ( arg == '-s' )
		launchClient = false
	else if ( arg == '-e' )
		host = ''
})

if ( ! musicPath ) {
	console.log( '\n\n\tMusic folder not defined.\n\tUse the command-line argument -m <path> to set the folder upon launching audioMotion.' )
	musicPath = readlineSync.questionPath(
		`\n\tPlease enter full path to music folder (e.g. ${ process.platform == 'win32' ? 'c:\\users\\john\\music' : '/home/john/music' })\n\tor just press Enter to use your home directory:\n\t> `, {
		isDirectory: true,
		defaultInput: '~'
	})
}

try {
	fs.accessSync( musicPath, fs.constants.R_OK ) // check if music folder is readable
}
catch (err) {
	showHelp()
	console.error( `\n\nERROR: no access to music folder at ${musicPath}` )
	process.exit(0)
}

// start Express web server
const server = express()
server.use( express.static( path.join( __dirname, '../public' ) ) )

server.use( '/music', express.static( musicPath ), ( req, res ) => {

	let files = getDir( musicPath + decodeURI( req.url ).replace( /%23/g, '#' ) ),
		imgs  = [];

	if ( files === false )
		res.status(404).send( 'Not found!' );
	else {
		files.files = files.files.filter( file => {
			if ( file.match( /\.(jpg|jpeg|png|gif|bmp)$/i ) )
				imgs.push( file );
			return ( file.match( /\.(mp3|flac|m4a|aac|ogg|wav|m3u|m3u8)$/i ) !== null );
		});
		files.cover = findImg( imgs, 'cover' ) || findImg( imgs, 'folder' ) || findImg( imgs, 'front' ) || imgs[0];
		res.send( files );
	}
})

server.listen( port, host, () => {
	console.log( `\n\n\tListening on port ${port} ${ host ? 'for localhost connections only' : 'accepting external connections!' }` )
	console.log( `\n\t/music mounted to ${musicPath}` )
	if ( launchClient ) {
		open( `http://localhost:${port}` )
		console.log( '\n\tLaunching client in browser...' )
	}
	console.log( '\n\nPress Ctrl+C to exit.' )
})

// route for custom server detection
server.get( '/serverInfo', ( req, res ) => {
	res.send( serverSignature )
})

// route for retrieving a directory's cover picture
server.get( '/getCover/:path', ( req, res ) => {
	const path    = musicPath + decodeURI( req.params.path ).replace( /^\/music/, '' ).replace( /%23/g, '#' ),
		  entries = getDir( path );

	if ( entries === false )
		res.status(404).send( 'Not found!' );
	else {
		const imgs = entries.files.filter( file => file.match( /\.(jpg|jpeg|png|gif|bmp)$/i ) !== null );
		res.send( findImg( imgs, 'cover' ) || findImg( imgs, 'folder' ) || findImg( imgs, 'front' ) || imgs[0] );
	}
});
